### DISCUSSION TITLE
Type casting a generic Protocol in overloaded decorators

### DISCUSSION THREAD

**Discussion by jameshclrk** [microsoft/pyright#3649] [Q&A]
Hello,

In case this is a case of the [XY problem](https://xyproblem.info/), what I'm trying to do:
* decorate a function
* return an object that:
    * "wraps" the function while maintaining the function's signature (`__call__` method)
    * has it's own attributes

and I would like the typing information in editors to show both the correct signature and the new attributes.

I have a solution, which is shown below, that uses a generic Protocol and I cast my object to that protocol. I *think* I'm not abusing the type system (eek), but this gives the correct result in editor completions and type info when hovering when /using/ the decorators, like in VS Code.

However, when I run `pyright` on the implementation, I get this error:

```
/private/tmp/hello/test_case.py:21:5 - error: Overloaded implementation is not consistent with signature of overload 2
    Function return type "(T@decorator) -> Proto[T@decorator]" is incompatible with type "((a: T@decorator) -> Proto[T@decorator]) | Proto[T@decorator]"
      Type "(T@decorator) -> Proto[T@decorator]" cannot be assigned to type "((a: T@decorator) -> Proto[T@decorator]) | Proto[T@decorator]"
        Type "(T@decorator) -> Proto[T@decorator]" cannot be assigned to type "(a: T@decorator) -> Proto[T@decorator]"
          Position-only parameter mismatch; expected 1 but received 0
        "object" is incompatible with protocol "Proto[T@decorator]"
          "__call__" is not present
          "some_other_attr" is not present (reportGeneralTypeIssues)
``` 
I'm looking for a sanity check that what I have done isn't too crazy and if this is a gap in `pyright`'s checker. Note: I've checked `mypy` and it doesn't display an error, but I'm not confident that it even tried to check the overloads.

Example code:
```python
from typing import Callable, Optional, Protocol, TypeVar, cast, overload

T = TypeVar("T")


class Proto(Protocol[T]):
    __call__: T
    some_other_attr: str


class MyObject:
    def __init__(self, fn):
        self.fn = fn

    def __call__(self, *args, **kwargs):
        return self.fn(*args, **kwargs)

    @property
    def some_other_attr(self):
        return "something"


@overload
def decorator(a: T) -> Proto[T]:
    ...


@overload
def decorator() -> Callable[[T], Proto[T]]:
    ...


def decorator(a: Optional[T] = None):
    def _inner(a: T):
        return cast(Proto[T], MyObject(a))

    if a is None:
        return _inner
    else:
        return _inner(a)
```

**Comment by erictraut** [comment]
You haven't provided a return type annotation for the implementation of `decorator`, so pyright is attempting to infer its type based on the returned values. The inferred return type isn't compatible with the declared return types of the overloads, which is why you're seeing this error. The solution is to provide a return type annotation on the `decorator` implementation, like this:

```python
def decorator(a: Optional[T] = None) -> Proto[T] | Callable[[T], Proto[T]]:
    ...
```

Alternatively, pyright will infer the correct return value if you make the decorator's input parameter positional-only, like this:
```python
def decorator(a: Optional[T] = None, /):
    ...
```

And yes, you're abusing the type system a bit here. Any time you need to use a `cast`, you should consider whether there's a better solution. In this, case, I would suggest using a `ParamSpec`. Also, your protocol class indicates that `some_other_attr` is a writable attribute, but your `MyObject` class implements a read-only property. I would stick with either a property or an attribute, not try to mix the two.

```python
from typing import Callable, Final, Generic, ParamSpec, Protocol, TypeVar, overload

_P = ParamSpec("_P")
_R = TypeVar("_R", covariant=True)

class Proto(Protocol[_P, _R]):
    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _R:
        ...

    some_other_attr: Final[str] = ""

class MyObject(Generic[_P, _R]):
    def __init__(self, fn: Callable[_P, _R]):
        self.fn = fn

    def __call__(self, *args: _P.args, **kwargs: _P.kwargs):
        return self.fn(*args, **kwargs)

    some_other_attr: Final[str] = "something"

@overload
def decorator(a: Callable[_P, _R]) -> Proto[_P, _R]:
    ...

@overload
def decorator() -> Callable[[Callable[_P, _R]], Proto[_P, _R]]:
    ...

def decorator(a: Callable[_P, _R] | None = None):
    def _inner(a: Callable[_P, _R], /) -> Proto[_P, _R]:
        return MyObject(a)

    if a is None:
        return _inner
    else:
        return _inner(a)
```
